### DISCUSSION TITLE
Adding line numbering to PDF and DOCX

### DISCUSSION THREAD

**Discussion by lorenzoFabbri** [quarto-dev/quarto-cli#6772] [Q&A] [answered]
### Description

I am writing a scientific manuscript with Quarto (version 1.2) and I need to export it to both PDF and DOCX. One of the requirements is line numbering.

For **PDF** I tried:

```
format:
  docx:
    reference-doc: reference_document.docx
  pdf:
    pdf-engine: lualatex
    number-sections: true
    toc: false
    include-in-header:
      text: |
        \usepackage{rotating}
        \usepackage{float}
        \usepackage{typearea}
        '\usepackage{lineno}\linenumbers'
```

which gives the following error message:

```
! LaTeX Error: Missing \begin{document}.

See the LaTeX manual or LaTeX Companion for explanation.
Type  H <return>  for immediate help.
 ...                                              
                                                  
l.107 '
     \usepackage{lineno}\linenumbers'
```

I also tried to use just `\usepackage{lineno}\linenumbers` (without the `'`) and putting `\linenumbers` in a new line, which give the error `[\directlua]:1: unexpected symbol near '\'.`. What partially worked was to put `\linenumbers` in a `latex` code chunk after the YAML, but this causes the abstract (which I keep in the YAML) not to be considered for line numbering.

I am totally unsure how to add line numbering to **DOCX**.

**Comment by mcanouil** [comment]
See the `lineno` documentation on how to use it.
`\linenumbers` needs to be in the body, not in the header.

````qmd
---
format:
  pdf:
    pdf-engine: lualatex
    number-sections: true
    toc: false
    keep-tex: true
    include-in-header:
      text: |
        \usepackage{rotating}
        \usepackage{float}
        \usepackage{typearea}
        \usepackage{lineno}
    include-before-body:
      text: |
        \linenumbers
---

{{< lipsum 4 >}}

````

**Reply by lorenzoFabbri** [reply]
Thank you but the abstarct is still without line numbering...

**Reply by mcanouil** [reply]
It's your LaTeX code not mine.
I simply fixed the usage.

**Reply by cscheid** [reply, marked as answer]
I think that what @mcanouil is saying is that since these two things are true at the same time:

- `lineno` only supports counting lines inside the LaTeX's document body
- `abstract` needs to be outside LaTeX's document body

Then you can't get line numbering for the abstract using `lineno`. Since you're using `lineno` yourself, there's nothing quarto can really do to help there.

**Reply by lorenzoFabbri** [reply]
Thank you for explaining! I temporarily fixed it by adding an abstract inside the body of the document...

**Reply by joelostblom** [reply]
Using `header-includes` instead of `include-in-header` works for me for numbering the abstract, titles, and paragraph text:

```yaml
pdf:
    header-includes:
      - \usepackage{lineno}
      - \linenumbers
```

![image](https://github.com/quarto-dev/quarto-cli/assets/4560057/e460a922-a3c5-446b-922c-07e928b4aa40)

**Reply by cderv** [reply]
Regarding lineo usage, `\linenumbers` is the way to turn on line numbering for the current mode. 

So you can add it in several places where you need to opt in line numbers. 

> Using header-includes instead of include-in-header works for me for numbering the abstract, titles, and paragraph text:

@joelostblom Just to clarify, using `include-in-header` works too 
````markdown
---
title: Hello
format:
  pdf:
    include-in-header:
      text: |
        \usepackage{lineno}
        \linenumbers
    abstract: |
      Some content
---

{{< lipsum 4 >}}

````
![image](https://github.com/quarto-dev/quarto-cli/assets/6791940/c3773352-badf-4545-9e60-2972b2a1bcea)

If only the abstract needs to be counted, it can be set in there
````markdown
---
title: Hello
format:
  pdf:
    include-in-header:
      text: |
        \usepackage{lineno}
    abstract: |
      \linenumbers
      Some content
---

{{< lipsum 4 >}}

````

![image](https://github.com/quarto-dev/quarto-cli/assets/6791940/e9452536-9570-4b74-b383-f4e167b3c6cc)

But you can also number abstract and text, without the title as shown above
````markdown
---
title: Hello
format:
  pdf:
    include-in-header:
      text: |
        \usepackage{lineno}
    abstract: |
      \linenumbers
      Some content
    include-before-body: 
      text: |
        \linenumbers
---

{{< lipsum 4 >}}

````
![image](https://github.com/quarto-dev/quarto-cli/assets/6791940/06d68dd6-7256-4fb9-af81-70cfd472792c)

So not sure what was not working @lorenzoFabbri 


Regarding Docx, you need to look in how it works in WORD. It is a layout option: https://support.microsoft.com/en-us/office/add-or-remove-line-numbers-b67cd35e-422c-42eb-adc9-256ca9802e22

Nothing to do with Quarto, as with the LaTeX solution. it is format specific
